system/jlib/jsocket.cpp

```diff
@@ -2116,26 +2122,32 @@ bool CSocket::check_connection()
     if (state != ss_open) 
         return false;
 
-    unsigned retrycount=100;
-EintrRetry:
-    int rc;
-    if (sockmode==sm_udp_server) { // udp server
-        DEFINE_SOCKADDR(u);
-        socklen_t  ul = setSockAddr(u,returnep,returnep.port);
-        rc = sendto(sock, NULL, 0, 0, &u.sa, ul);
-    }
-    else {
-        rc = send(sock, NULL, 0, SEND_FLAGS);
-    }
-    if (rc < 0) {
+    int rc = wait_read(0);
+    if (rc == 0)
+        return true;
+    else if (rc < 0)
+        return false;
+
+    int retrycount = 100;
+    char buffer;
+
+EintrRecv:
+    rc = recv(sock, &buffer, sizeof(buffer), MSG_PEEK);
+
+    if (rc == 0)
+        return false;
```

Comment:
not sure I understand this, can't the connection be good, but 0 bytes on the wire?
from recv man:
```
       When a stream socket peer has performed an orderly shutdown, the
       return value will be 0 (the traditional "end-of-file" return).

       Datagram sockets in various domains (e.g., the UNIX and Internet
       domains) permit zero-length datagrams.  When such a datagram is
       received, the return value is 0.

       The value 0 may also be returned if the requested number of bytes
       to receive from a stream socket was 0.
```

Reply:
For a UDP socket it could be triggered by someone sending a 0-byte datagram, but not sure this code is supposed to handle this. For TCP it should be ok.

Reply:
Is it ok, because wait_read(0) would have return 0, which usually means timeout, but because poll timeout is 0, means 0 file descriptors ready/0 errors, so has returned true before got here...
So only here because there should have been some date to read, and if not = error   ?
Could benefit from a comment describing semantics.

I guess this is knowingly not thread safe? i.e. if there was data on the socket and got to the recv, but someone else read it before this recv did, then socket is good, but this would return false.
Probably ok, assumes not multiple threads reading/calling check_connection concurrently ?

Reply:
Doesn't matter what the poll timeout is - wait_read() returning 0 means socket is not closed at that moment.
If wait_read() returns > 0 then we recv() knowning it wont block - and if that returns 0 then we know socket was closed.

Reply:
I'll add a comment

Reply:
makes sense now, thanks.